src/pip/_internal/vcs/helix_core.py

```diff
@@ -0,0 +1,286 @@
+from __future__ import absolute_import
+
+import logging
+import os
+import random
+import re
+import socket
+import string
+
+from pip._vendor.six.moves.urllib import parse as urllib_parse
+from pip._internal.vcs import VersionControl, vcs
+
+logger = logging.getLogger(__name__)
+
+client_spec_template = """
+Owner:   {user}
+Host:    {host}
+Client:  {name}
+Root:    {client_path}
+View:    /{server_path}/... //{name}/...
+Options: allwrite rmdir
+"""
+
+
+def split_url(url, environ=os.environ):
+    """
+    Parses a URL and returns a 5-tuple of
+    ``(port, depot_path, revision, username, password)``. Where fields are not
+    present in the given URL, information from Helix Core environment variables
+    (``P4PORT``, ``P4USER``, ``P4PASSWD``) is used.
+
+    URL examples::
+
+        ssl://me:p4ssw0rd@perforce1/depot/myproject
+        p4://perforce1:1667/depot/myproject
+        p4:///depot/myproject
+    """
+
+    if url.startswith("p4+"):
+        url = url[3:]
+    url = urllib_parse.urlsplit(url)
+
+    p4path = url.path
+    if '@' in p4path:
+        p4path, p4rev = p4path.split("@")
+    else:
+        p4rev = None
+
+    p4user = (
+        url.username or
+        environ.get("P4USER") or
+        environ.get("USER") or
+        environ.get("USERNAME") or
+        '')
```

Comment:
I would define `get_p4user()` and `get_p4password()` functions.

Reply:
Could you explain why? The code here would be no more self-contained if moved to another function, as it doesn't use any local variables. I could add some sort of `underlay_environ()` function and call that repeatedly, but sadly `P4PORT` can use a number of formats (`port`, `host:port` or `protocol:host:port`).

Reply:
It's more testable that way since you can test the logic independent of the surrounding function. It also reduces the cyclotomic complexity of the surrounding function, which already has a lot of branching.